Add NativeMicrophone feature to improve audio quality

leejeonghun opened the pull request:
기존 유니티 Microphone 에서 생성한 AudioClip 에서 오디오 데이터를 가져와 DummyAudioDevice 를 통해 전달하는 방식은 유니티 쓰레드와의 호출 주기 차이에 대응하기 위한 버퍼링(지연증가) 및 정밀하지 못한 오디오 타이밍 처리 때문에 여러가지 품질 저하의 원인이 되어 왔습니다.

지금까지 여러가지 방법을 모색해 가며 대응해 왔으나, 보다 근본적인 해결 및 추구하는 품질 기준을 확실히 만족시키기 위해, libwebrtc 에서 제공하는 오디오 파이프라인 전체를 그대로 활용할 수 있도록 마이크 관련 API(마이크 목록 확인 및 선택)를 추가하고,  libwebrtc 자체 오디오 재생 기능을 비활성화 하여 유니티에서 libwebrtc 의 오디오 품질 그대로 활용할 수 있도록 수정 하였습니다.

다양하게 테스트 해 본 결과, 12월 빌드에 걸맞는 오디오 품질에 도달했다 판단되어 PR 올립니다.
아, 인터페이스 단 breaking change 가 있어 클라이언트가 master 가 아니라 본 브렌치를 보도록 우선 처리 후 머지되면 master 를 보도록 수정할 예정이오니 리뷰 후 merge 는 제가 클라이언트와 타이밍을 맞춰 진행할 예정입니다. 참고 부탁 드리겠습니다.

leejeonghun committed:
Add NativeMicrophone feature to improve audio quality

leejeonghun committed:
Remove no longer used files

heonbert commented:
+_+ 수고하셨습니다!
얼른 테스트를 해보고 싶네요.

그리고 이제 Unity의 AudioClip을 거치지 않는 만큼, 데이터 조작을 통한 음성 변조 등을 하고 싶다면 어떻게 해야 할지도 생각해봐야겠네요.

leejeonghun committed:
Add missing files

leejeonghun commented:
그 부분도 고려했고, 클라쪽 PR 에 언급할 예정입니다 :) 걱정마시길!

leejeonghun committed:
Add channel count correction

leejeonghun committed:
Activate debug information for dump analysis

leejeonghun committed:
Add null check to TransceiverStop()

heonbert commented on Runtime/Scripts/AudioStreamTrack.cs:
```diff
@@ -274,6 +260,12 @@ private void OnAudioReceivedInternal(float[] audioData, int sampleRate, int chan
                     frameCountReceiveDataForIgnoring++;
                     return;
                 }
+                if (audioData.Length != channels * numOfFrames)
```
오오! 이렇게 실제 채널 수를 체크할 수 있군요.

heonbert reviewed (approved):
수고하셨습니다. ^^

leejeonghun committed:
Add a workaround when input audio is as single channel

heonbert reviewed (approved):
강제 2채널이군요. 😂 일단 테스트는 가능해지겠네요! 감사합니다.

leejeonghun committed:
Revert a forced stereo workaround

leejeonghun committed:
Add OnAudioConfigurationChanged to recreate AudioClip when device is lost

leejeonghun committed:
Fix incorrect patch context

leejeonghun committed:
Add webrtc.dll and webrtc.pdb with debug symbols

Public forks are not support LFS, so pdb file is compressed using zip to avoid Github file size limit.

leejeonghun commented:
PR을 여러개 올리기 그래서 본 PR에 FP-560 과 FP-743 에 대한 commit 을 포함시켰습니다. 자세한 내용은 bitbucket 쪽 PR 에 작성하겠습니다.

heonbert reviewed (approved):
수고하셨습니다👍

leejeonghun committed:
Rename to be ignored by Unity
